src/common/schedulers/slurm_commands.py

```diff
@@ -440,3 +457,49 @@ def _parse_nodes_info(slurm_node_info):
         if lines:
             slurm_nodes.append(SlurmNode(*lines))
     return slurm_nodes
+
+
+def get_node_type(nodename):
+    """Get node type given nodename. Node type format: {queue_name}-{instancetype}."""
+    queue_name, _, instance_name = parse_nodename(nodename)
+    return queue_name + "-" + instance_name
+
+
+def get_nodes_types_from_slurm_nodes(nodes):
+    """Get node type set given Slurm nodes. Node type format: {queue_name}-{instancetype}."""
+    nodes_types = set()
+    for node in nodes:
+        nodes_types.add(get_node_type(node.name))
+    return nodes_types
+
+
+def get_inactive_partitions_name(partitions):
+    """Get the name of all inactive partitions."""
+    inactive_partitions_name = []
+    for partition_name, partition in partitions.items():
+        if partition.state == "INACTIVE":
+            inactive_partitions_name.append(partition_name)
+    return inactive_partitions_name
+
+
+def get_partitions_with_active_jobs(nodename_to_slurm_nodes_map, partitions):
+    """Retrieve partitions currently have active jobs running."""
+    partitions_have_active_jobs = set()
+    for partition_name, partition in partitions.items():
+        for node_name in partition.nodes.split(","):
+            if (
+                nodename_to_slurm_nodes_map.get(node_name)
+                and nodename_to_slurm_nodes_map.get(node_name).is_running_job()
+            ):
+                partitions_have_active_jobs.add(partition_name)
+                break
+    return partitions_have_active_jobs
+
+
+def retrieve_partitions_from_node_types(node_types):
+    """Retrieve partitions for node types."""
+    successful_launched_partitions = []
+    for nodes_type in node_types:
+        partitions, _ = re.match(r"^([a-z0-9\-]+)-([a-z0-9]+)", nodes_type).groups()
```

Comment:
As we discussed, please try to avoid parsing information we should already have from `parse_nodename` again